Move documentation from the README to odoc pages

gpetiot opened the pull request:
Even if this PR is not finished, any feedback and help is of course welcomed!

- [x] Trimming the README to only keep a "getting started" section and link to the full manual
- [x] Gather the questions in a FAQ page (currently in the wiki)
- [x] Mention `dune` integration (I can't believe we didn't)
- [x] Mention `merge-fmt` in the Workflows page (fix #1155)
- [x] Document the doc-comment language accepted (fix #1347)

cc @tmattio

tmattio commented on doc/getting_started.mld:
```diff
@@ -0,0 +1,78 @@
+ocamlformat [OPTION]... [SRC]...
+]}
+
+The most common usecase involves using the Dune build system, once your project is correctly setup (see {{:https://dune.readthedocs.io/en/stable/formatting.html#formatting-a-project}Dune's manual}) you can reformat your project using:
```
Would be cool to have some explanation on how it works. For instance:

- Does `fmt` only work with `ocamlformat`? (I think it works with `refmt` for Reason)
- Which file does it format? (only the ones in that have a dune rule I think)
- etc.

gpetiot replied:
done

tmattio commented on doc/getting_started.mld:
```diff
@@ -0,0 +1,78 @@
+
+{2 Code style}
+
+There are a number of preset code style profiles, selected using the `--profile` option by passing `--profile=<name>` on the command line or adding `profile = <name>` to an .ocamlformat configuration file. Each profile is a collection of settings for all options, overriding lower priority configuration of individual options. So a profile can be selected and then individual options can be overridd ...
```
We should maybe add a list of profiles somewhere

gpetiot replied:
done

tmattio commented on doc/getting_started.mld:
```diff
@@ -0,0 +1,78 @@
+
+`.ocamlformat` files in the containing and all ancestor directories for each input file are used, as well as the global .ocamlformat file defined in `$XDG_CONFIG_HOME/ocamlformat`. The global .ocamlformat file has the lowest priority, then the closer the directory is to the processed file, the higher the priority.
+
+When the option `--enable-outside-detected-project` is not set, .ocamlformat files outside of the project (including the one in `XDG_CONFIG_HOME`) are not read. The project root of an input file is taken to be the nearest ancestor directory that contains a .git or .hg or dune-project file. If no config file is found, formatting is disabled.
```
Having this sentence in a negative form (when it's **not** set) makes it a bit confusing.

gpetiot replied:
I agree, done.

tmattio commented on doc/workflows.mld:
```diff
@@ -0,0 +1,50 @@
+{0 OCamlFormat Manual}
+
+{1 Workflows}
```
I'd rename this to How-Tos and add a small description of the description. It wasn't obvious what the goal was until I read a couple of sub-sections.

gpetiot replied:
done

tmattio commented on doc/editor_setup.mld:
```diff
@@ -0,0 +1,92 @@
+
+As mentioned in the Options section, when the option `--disable-outside-detected-project` is set, .ocamlformat files outside of the project (including the one in `XDG_CONFIG_HOME`) are not read. The project root of an input file is taken to be the nearest ancestor directory that contains a .git or .hg or dune-project file. If no config file is found, then the formatting is disabled.
+
+This feature is often the behavior you can expect from OCamlFormat when it is directly run from your text editor, so it is advised to use this option.
```
I'd start this section with "Make sure you have a `.ocamlformat` file at the root of your project". We frequently have issues in VSCode for this, so it's good to surface the need to create this file in this Editor Setup doc of ocamlformat 🙂

gpetiot replied:
done

tmattio commented on doc/faq.mld:
```diff
@@ -0,0 +1,91 @@
+
+{2 How to ignore directories?}
+
+It is possible to disable OCamlFormat for the files of a directory by having an `.ocamlformat` file containing disable in this directory, or listing the files to ignore in an `.ocamlformat-ignore` file.
```
Would be nice if it worked recursively by the way 😄 In the meantime, maybe a quick note that you need to list all the descendants of the directory to ignore them.

gpetiot replied:
done

tmattio reviewed (commented):
This is great, thank you so much for working on this!

I left some suggestions, but feel free to ping me when you're done editing for a second pass of review 🙂

gpetiot commented on doc/getting_started.mld:
```diff
@@ -0,0 +1,78 @@
+Alternatively, it can be built manually with:
+
+{[
+dune subst
```
It's to set ocamlformat's version, I will add a comment in the cmd

gpetiot commented:
@Julow @jonludlam @tmattio I would like to have 3 pages that would present the styles of each profile (show how functions, let bindings, etc. are formatted). These would be basically three copies of the same file but setting `ocamlformat` to a different configuration.

I don't know if there is plans to integrate `ocamlformat` into `odoc` to automatically reformat ocaml code blocks (should probably use `ocamlformat-rpc`), if not we could support .mld files in ocamlformat, we would need more functions from the `odoc-parser` API to read a file and pretty print most of the elements so we could override the pretty printing of code blocks.

Julow reviewed (commented):
> we could support .mld files in ocamlformat

This is a good idea !
`odoc-parser` can already parse mld files. I think ocamlformat should take care of formatting the whole file, as it does for comments.

I think this is better done in ocamlformat rather than in odoc, as ocamlformat already formats doc-comments and already depends on odoc-parser.

gpetiot commented on doc/howtos.mld:
```diff
@@ -0,0 +1,51 @@
+
+{2 Resolve merge conflicts using Merge-fmt and OCamlFormat}
+
+{{:https://github.com/hhugo/merge-fmt}Merge-fmt} is a small wrapper on top of [git] commands to help resolve conflicts by leveraging code formatters.
```
done

gpetiot commented on doc/index.mld:
```diff
@@ -0,0 +1,18 @@
+OCamlFormat is probably what you are after!
+OCamlFormat is a tool to format OCaml code.
+
+OCamlFormat works by parsing source code using the OCaml compiler's standard parser, deciding where to place comments in the parse tree, and writing the parse tree and comments in a consistent style.
```
done

gpetiot commented:
I updated the PR with the review comments, the part about giving some examples for each profile requires a tool that would be able to format .mld files. I will have a look at it in another PR, but it shouldn't delay this one, so I removed theses pages for now.
@tmattio it's ready for a second round of review if you hvae time.

Where to deplay these pages can be taken care of later, but the generated pages should automatically be uploaded on https://v3.ocaml.org/p/ocamlformat/

gpetiot commented:
Any last reviews on this? and any insight about where to host this?

Julow commented on doc/faq.mld:
```diff
@@ -0,0 +1,103 @@
+
+OCamlFormat is influenced by and follows the same basic design as [refmt] for {{:https://github.com/facebook/reason}Reason}, but outputs OCaml instead of Reason.
+
+This tool is not able to deal directly with Reason code ([*.re]/[*.rei] files),
+but it is possible to first convert these files to OCaml syntax using [refmt -p
+ml] and then running [ocamlformat] on this output.
```
This means that OCamlformat can be used to convert reason code into ocaml code, not "handle" reason files.

gpetiot replied:
done

Julow commented on doc/faq.mld:
```diff
@@ -0,0 +1,103 @@
+dir/*
+dir/*/*
+dir/*/*/*
+]}
```
Perhaps we can suggest to add a ignore file with just `*` in all directory that needs to be ignored ?

gpetiot replied:
done

Julow commented on doc/faq.mld:
```diff
@@ -0,0 +1,103 @@
+
+{2 How to locally disable OCamlFormat?}
+
+To disable the formatting of a specific toplevel item you must attach an [[@@ocamlformat "option=VAL"]] attribute to this item in the processed file, such as:
```
```suggestion
To disable the formatting of a specific toplevel item you must attach an [[@@ocamlformat "disable"]] attribute to this item in the processed file, such as:
```

Julow commented on doc/faq.mld:
```diff
@@ -0,0 +1,103 @@
+] [@@ocamlformat "disable"]
+]}
+
+To disable the formatting of a specific expression you must attach an [[@ocamlformat "option=VAL"]] attribute to this expression in the processed file, such as:
```
```suggestion
To disable the formatting of a specific expression you must attach an [[@ocamlformat "disable"]] attribute to this expression in the processed file, such as:
```

Julow commented on doc/faq.mld:
```diff
@@ -0,0 +1,103 @@
+
+{2 Why is there an inconsistent spacing on the | character in match blocks?}
+
+The change in the indentation of [| ] vs [|] would indicate that there is a short body of a nested or-pattern that could be easily missed if formatted as: [| ]. This can be disabled using [indicate-nested-or-patterns = false] if you are happy with the risk. Alternatively, [break-cases = nested] can be used to break after the [->] to avoid the risk at the cost of significantly less compact code.
```
```suggestion
The change in the indentation of [| ] vs [|] would indicate that there is a short body of a nested or-pattern that could be easily missed if formatted as: [| ]. This can be disabled using [indicate-nested-or-patterns = unsafe-no] if you are happy with the risk. Alternatively, [break-cases = nested] can be used to break after the [->] to avoid the risk at the cost of significantly less compact code.
```

The default is already `unsafe-no`, so I think it's reasonable to not say it's a risk when we already decided it is fine.

Julow commented on doc/index.mld:
```diff
@@ -0,0 +1,15 @@
+OCamlFormat is a tool to format OCaml code.
+
+{1 Table of Content}
+- {{!page-getting_started}Getting started}
+- {{!page-editor_setup}Editor setup}
+- {{!page-howtos}How-To's}
+- {{!page-faq}Frequently asked questions}
+- {{!page-doc_comments}Doc-comments language reference}
```
The odoc 2.0 way is:
```suggestion
- {{!childpage-getting_started}Getting started}
- {{!childpage-editor_setup}Editor setup}
- {{!childpage-howtos}How-To's}
- {{!childpage-faq}Frequently asked questions}
- {{!childpage-doc_comments}Doc-comments language reference}
```

jonludlam replied:
I think if you do that dune/odig will break - `page` would be best here.

Julow replied:
It worked for me, locally. I think both prefixes have the same behavior at the moment.

jonludlam replied:
The idea of `childpage` was to restrict lookup to only the direct children of the page, but since dune and odig have a flat namespace for the pages because they just compile with `--package`, they aren't children of `index.mld`.

Julow reviewed (commented):
This looks really nice !

gpetiot commented on README.md:
```diff
@@ -2,294 +2,62 @@
-
-* `ocamlformat --help`
-* `man ocamlformat`
+If you have further questions, please read OCamlFormat's user manual: <...>
```
That's supposed to be the link to these pages, probably https://v3.ocaml.org/p/ocamlformat/

gpetiot replied:
done

gpetiot commented on README.md:
```diff
@@ -2,294 +2,62 @@
-- an `[@@ocamlformat "option=VAL"]` attribute on an expression in the processed file
-
-.ocamlformat files in the containing and all ancestor directories for each input file are used, as well as the global .ocamlformat file defined in `$XDG_CONFIG_HOME/ocamlformat`. The global .ocamlformat file has the lowest priority, then the closer the directory is to the processed file, the higher the priority.
+OCamlFormat is a tool to format OCaml code.
```
done

gpetiot commented on doc/editor_setup.mld:
```diff
@@ -0,0 +1,94 @@
+            \ }
+
+let g:neoformat_enabled_ocaml = ['ocamlformat']
+]}
```
If this is not polished enough we can wait and open another PR when this is ready, we need to be careful when advertising options that are not part of the recommended config.

gpetiot commented on doc/editor_setup.mld:
```diff
@@ -0,0 +1,94 @@
+
+{1 Editor setup}
+
+First of all, make sure you have an [.ocamlformat] file at the root of your project.
```
done

gpetiot commented on doc/doc_comments.mld:
```diff
@@ -0,0 +1,156 @@
+{0 OCamlFormat Manual}
+
+{1 Doc-comments language reference}
```
I've written a small example, let me know what you think, maybe it's too much. Anyway there is now a link pointing to the odoc documentation.

gpetiot commented on README.md:
```diff
@@ -2,294 +2,62 @@
-
 Hello, new user! Welcome! :wave:
 
 If you are here, you are probably interested in using a formatting tool for your
 code base, so that you do not have to worry about formatting it by hand, and to
 speed up code review by focusing on the important parts.
```
done

Julow commented on doc/doc_comments.mld:
```diff
@@ -0,0 +1,42 @@
+{0 Doc-comments language reference}
+
+OCamlFormat uses {{:https://github.com/ocaml/odoc}odoc} to parse doc-comments (also referred to as doc-strings), and hence it inherits the accepted language from [odoc] (detailed in {{:https://ocaml.github.io/odoc/odoc_for_authors.html}odoc's documentation}).
```
```suggestion
OCamlFormat uses {{:https://github.com/ocaml-doc/odoc-parser}odoc-parser} to parse doc-comments (also referred to as doc-strings), and hence it inherits the accepted language from {{:https://github.com/ocaml/odoc}odoc} (detailed in {{:https://ocaml.github.io/odoc/odoc_for_authors.html}odoc's documentation}).
```

It moved :) I changed the second occurrence of `odoc` into a link to not loose it.

Julow commented on doc/doc_comments.mld:
```diff
@@ -0,0 +1,42 @@
+
+{%html:
+<pre>
+<code class="ml hljs ocaml">(** {0 Adding integers} *)
```
`{0 }` is only possible for pages, not signatures because the rendered documentation will have the equivalent of `{0 module Foo}` already at the top. The most reasonable thing to write here is a synopsis: a paragraph explaining the module:

```suggestion
<code class="ml hljs ocaml">(** Adding integers. *)
```

Julow commented on doc/doc_comments.mld:
```diff
@@ -0,0 +1,42 @@
+
+    Notes:
+    - {_ check} that exception {!exception:Int_overflow} is {b not raised};
+    - have a look at {!module:Int} ({{:https://ocaml.org/api/Int.html}Documentation}).
```
The link to the documentation is weird, the reference is already a better link to the same documentation because it stays on the same site. Why not a link to the RWO chapter about the encoding of ints: https://dev.realworldocaml.org/runtime-memory-layout.html#scrollNav-2-1

gpetiot replied:
The idea was to show the syntax for links, but yeah it doesn't have much and it's not very clear.

Julow reviewed (approved):
Looks good :)
